search.py

```diff
@@ -406,101 +406,92 @@ def astar_search(problem, h=None):
-        return index_blank_square
+        '''Return the index of the blank square in a given state'''
+
+        return state.index(0)
     
     def actions(self, state):
-        """Return the actions that can be executed in the given state.
+        ''' Return the actions that can be executed in the given state.
         The result would be a list, since there are only four possible actions
-        in any given state of the environment."""
-       
-        possible_actions = list()
+        in any given state of the environment '''
+        
+        possible_actions = ['UP', 'DOWN', 'LEFT', 'RIGHT']       
         index_blank_square = self.find_blank_square(state)
 
-        if index_blank_square(0) == 0:
-            possible_actions += ['DOWN']
-        elif index_blank_square(0) == 1:
-            possible_actions += ['UP', 'DOWN']
-        elif index_blank_square(0) == 2:
-            possible_actions += ['UP']
-        
-        if index_blank_square(1) == 0:
-            possible_actions += ['RIGHT']
-        elif index_blank_square(1) == 1:
-            possible_actions += ['LEFT', 'RIGHT']
-        elif index_blank_square(1) == 2:
-            possible_actions += ['LEFT']
+        if index_blank_square % 3 == 0:
+            possible_actions.remove('LEFT')
+        if index_blank_square < 3:
+            possible_actions.remove('UP')
+        if index_blank_square % 3 == 2:
+            possible_actions.remove('RIGHT')
+        if index_blank_square > 5:
+            possible_actions.remove('DOWN')
 
         return possible_actions
 
     def result(self, state, action):
-        """Given state and action, return a new state that is the result of the action.
-        Action is assumed to be a valid action in the state."""
-
-        blank_square = self.find_blank_square(state)
-        new_state = [row[:] for row in state]
-
-        if action=='UP':
-            new_state[blank_square(0)][blank_square(1)] = new_state[blank_square(0)-1][blank_square(1)]
-            new_state[blank_square(0)-1][blank_square(1)] = 0
-        elif action=='LEFT':
-            new_state[blank_square(0)][blank_square(1)] = new_state[blank_square(0)][blank_square(1)-1]
-            new_state[blank_square(0)][blank_square(1)-1] = 0
-        elif action=='DOWN':
-            new_state[blank_square(0)][blank_square(1)] = new_state[blank_square(0)+1][blank_square(1)]
-            new_state[blank_square(0)+1][blank_square(1)] = 0
-        elif action=='RIGHT':
-            new_state[blank_square(0)][blank_square(1)] = new_state[blank_square(0)][blank_square(1)+1]
-            new_state[blank_square(0)][blank_square(1)+1] = 0
+        ''' Given state and action, return a new state that is the result of the action.
+        Action is assumed to be a valid action in the state '''
+
+        # ix is the index of the blank square
+        ix = self.find_blank_square(state)
+        new_state = list(state)
+
+        if action == 'UP':
```

Comment:
These 8 lines could be

      delta = dict(UP=-3, DOWN=3, LEFT=-1, RIGHT=1)
      neighbor = blank + delta[action]
      new_state[blank], new_state[neighbor] = new_state[neighbor], new_state[blank]

Reply:
This is clever. I don't know why I didn't think of this.